File: native_locator/src/known.rs
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.
use std::{env, path::PathBuf};

pub trait Environment {
    fn get_user_home(&self) -> Option<PathBuf>;
    /**
     * Only used in tests, this is the root `/`.
     */
    #[allow(dead_code)]
    fn get_root(&self) -> Option<PathBuf>;
    fn get_env_var(&self, key: String) -> Option<String>;
    fn get_know_global_search_locations(&self) -> Vec<PathBuf>;
}

pub struct EnvironmentApi {}
impl EnvironmentApi {
    pub fn new() -> Self {
        EnvironmentApi {}
    }
}

#[cfg(windows)]
impl Environment for EnvironmentApi {
    fn get_user_home(&self) -> Option<PathBuf> {
        get_user_home()
    }
    fn get_root(&self) -> Option<PathBuf> {
        None
    }
    fn get_env_var(&self, key: String) -> Option<String> {
        get_env_var(key)
    }
    fn get_know_global_search_locations(&self) -> Vec<PathBuf> {
        vec![]
    }
}

#[cfg(unix)]
impl Environment for EnvironmentApi {
    fn get_user_home(&self) -> Option<PathBuf> {
        get_user_home()
    }
    fn get_root(&self) -> Option<PathBuf> {
        None
    }
    fn get_env_var(&self, key: String) -> Option<String> {
        get_env_var(key)
    }
    fn get_know_global_search_locations(&self) -> Vec<PathBuf> {
        vec![
            PathBuf::from("/usr/bin"),
            PathBuf::from("/usr/local/bin"),
            PathBuf::from("/bin"),
            PathBuf::from("/home/bin"),
            PathBuf::from("/sbin"),
            PathBuf::from("/usr/sbin"),
            PathBuf::from("/usr/local/sbin"),
            PathBuf::from("/home/sbin"),
            PathBuf::from("/opt"),
            PathBuf::from("/opt/bin"),
            PathBuf::from("/opt/sbin"),
            PathBuf::from("/opt/homebrew/bin"),
        ]
    }
}

fn get_user_home() -> Option<PathBuf> {
    let home = env::var("HOME").or_else(|_| env::var("USERPROFILE"));
    match home {
        Ok(home) => Some(PathBuf::from(home)),
        Err(_) => None,
    }
}

fn get_env_var(key: String) -> Option<String> {
    match env::var(key) {
        Ok(path) => Some(path),
        Err(_) => None,
    }
}

File: native_locator/ts-sync/src/known.ts
// import { env } from 'process';
// import { PathLike } from 'fs';

// export interface Environment {
//     get_user_home(): string | undefined;
//     get_env_var(key: string): string | undefined;
//     get_known_global_search_locations(): PathLike[];
// }

// export class EnvironmentApi implements Environment {
//     get_user_home(): PathLike | undefined {
//         return get_user_home();
//     }

//     get_env_var(key: string): string | undefined {
//         return get_env_var(key);
//     }

//     get_known_global_search_locations(): PathLike[] {
//         return [];
//     }
// }

// function get_user_home(): PathLike | undefined {
//     const home = env.HOME || env.USERPROFILE;
//     return home ? home : undefined;
// }

// function get_env_var(key: string): string | undefined {
//     return env[key];
// }

export function getKnowGlobalSearchLocations(): string[] {
    return [
        '/usr/bin',
        '/usr/local/bin',
        '/bin',
        '/home/bin',
        '/sbin',
        '/usr/sbin',
        '/usr/local/sbin',
        '/home/sbin',
        '/opt',
        '/opt/bin',
        '/opt/sbin',
        '/opt/homebrew/bin',
    ];
}
